tools/shared.py

```diff
@@ -1637,35 +1604,94 @@ def closure_compiler(filename, pretty=True):
   @staticmethod
   def is_ar(filename):
     try:
-      if Building._is_ar_cache.get(filename):
+      if filename in Building._is_ar_cache:
         return Building._is_ar_cache[filename]
-      b = open(filename, 'r').read(8)
-      sigcheck = b[0] == '!' and b[1] == '<' and \
-                 b[2] == 'a' and b[3] == 'r' and \
-                 b[4] == 'c' and b[5] == 'h' and \
-                 b[6] == '>' and ord(b[7]) == 10
+      with open(filename, 'r') as f:
+        b = f.read(8)
+        sigcheck = len(b) == 8 and \
+                   b[0] == '!' and b[1] == '<' and \
+                   b[2] == 'a' and b[3] == 'r' and \
+                   b[4] == 'c' and b[5] == 'h' and \
+                   b[6] == '>' and ord(b[7]) == 10
       Building._is_ar_cache[filename] = sigcheck
       return sigcheck
     except Exception, e:
       logging.debug('Building.is_ar failed to test whether file \'%s\' is a llvm archive file! Failed on exception: %s' % (filename, e))
       return False
 
+  _is_bitcode_cache = {}
   @staticmethod
   def is_bitcode(filename):
-    # look for magic signature
-    b = open(filename, 'r').read(4)
-    if b[0] == 'B' and b[1] == 'C':
-      return True
-    # look for ar signature
-    elif Building.is_ar(filename):
-      return True
-    # on OS X, there is a 20-byte prefix
-    elif ord(b[0]) == 222 and ord(b[1]) == 192 and ord(b[2]) == 23 and ord(b[3]) == 11:
-      b = open(filename, 'r').read(24)
-      return b[20] == 'B' and b[21] == 'C'
+    try:
+      if filename in Building._is_bitcode_cache:
```

Comment:
OK, I've done some perf testing on the three caches that are involved in these commits. Some slightly less scientific testing than the runs I did above, just to see if the caching has any effect at all. I didn't do as many repeats for each option here.

### High-level benchmarking
#### 300 sources compile

option | time
-------|-----
all caches present | 2:22:3
all caches removed | 2:21.7 (17.3 system)
is_ar cache only | 2.22:8 (17.4 system)
is_bitcode cache only | 2:23.5 (17.6 system)
ar_contents cache only | 2:21.4 (16.9 system)

#### Link a single output file

option | time
-------|-----
all caches present | 14.5 (1.4 system)
all caches removed | 14.4 (1.4 system)
is_ar cache only | 14.3 (1.4 system)
is_bitcode cache only | 14.3 (1.3 system)
ar_contents cache only | 14.4 (1.4 system)

#### Conclusion

Hard to be sure from the high-level overall timings, the margins are so small! The tables above are basically useless, the effect of the caches is completely swamped by noise.

### Some more micro-benchmarking of the specific functions

It appears that the `ar_contents` cache is useful, giving a consistent 200ms reduction in disc access time (system time). It saves calling LLVM `ar` more than half the time during the link stage. I'd keep this one in.

The `is_ar` cache is useful too I think; over half the time it kicks in and prevents a disk access (specifically, around 40 cache hits, 30 cache misses for typical link lines throughout our projects). I can't measure any timing impact of that at all though!

The `is_bitcode` cache is completely useless, each bitcode file is only queried once for all the compilation and link commands in our project.